### PR TITLE
Fix _mm_test_mix_ones_zeros and _mm_testnzc_si128

### PR DESCRIPTION
**Pull request by aqrit**
Bug: `_mm_test_mix_ones_zeros` always returned true.
The function wasn't reducing `zf` and `cf` to a bool before combining them.

The fix proposed here isn't the most efficient, but at least it is correct. 

` _mm_testnzc_si128` is an alias for `_mm_test_mix_ones_zeros`.

Note(s):
The arguments are named incorrectly in the `_mm_test_mix_ones_zeros` documentation[0].
The second argument is the mask, as per the behavior of `_mm_test_mix_ones_zeros` with gcc and clang.
This naming error seems to have propagated through both gcc[1] and llvm[2] headers but not to rust[3] headers or sse2neon[4].

[0] https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=ptest&techs=SSE_ALL&ig_expand=6902
[1] https://github.com/gcc-mirror/gcc/blob/27ce74fa23c93c1189c301993cd19ea766e6bdb5/gcc/config/i386/smmintrin.h#L94
[2] https://github.com/llvm/llvm-project/blob/70535f5e609f747c28cfef699eefb84581b0aac0/clang/lib/Headers/smmintrin.h#L1130
[3] https://github.com/rust-lang/stdarch/blob/f4528dd6e85d97bb802240d7cd048b6e1bf72540/crates/core_arch/src/x86/sse41.rs#L1149
[4] https://github.com/DLTcollab/sse2neon/blob/243e90f654193c97a691b1a53213d091e02eb631/sse2neon.h#L7595

### CODE REVIEW

File: sse2neon.h [hunk 1]
```diff
@@ -7592,14 +7592,10 @@ FORCE_INLINE int _mm_test_all_zeros(__m128i a, __m128i mask)
...
+// Note: Argument names may be wrong in the Intel intrinsics guide.
 FORCE_INLINE int _mm_test_mix_ones_zeros(__m128i a, __m128i mask)
 {
-    uint64x2_t zf =
-        vandq_u64(vreinterpretq_u64_m128i(mask), vreinterpretq_u64_m128i(a));
-    uint64x2_t cf =
-        vbicq_u64(vreinterpretq_u64_m128i(mask), vreinterpretq_u64_m128i(a));
-    uint64x2_t result = vandq_u64(zf, cf);
-    return !(vgetq_lane_u64(result, 0) | vgetq_lane_u64(result, 1));
+    return !(_mm_testz_si128(a, mask) | _mm_testc_si128(a, mask));
```
Comment (aqrit) [inline sse2neon.h:7598]: Needs a forward declare, or should I just copy and paste these functions into this functions body?

File: sse2neon.h [hunk 2]
```diff
@@ -7592,15 +7592,8 @@ FORCE_INLINE int _mm_test_all_zeros(__m128i a, __m128i mask)
 // zero, otherwise set CF to 0. Return 1 if both the ZF and CF values are zero,
 // otherwise return 0.
 // https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=mm_test_mix_ones_zero
-FORCE_INLINE int _mm_test_mix_ones_zeros(__m128i a, __m128i mask)
-{
-    uint64x2_t zf =
-        vandq_u64(vreinterpretq_u64_m128i(mask), vreinterpretq_u64_m128i(a));
-    uint64x2_t cf =
-        vbicq_u64(vreinterpretq_u64_m128i(mask), vreinterpretq_u64_m128i(a));
-    uint64x2_t result = vandq_u64(zf, cf);
```
Comment (howjmay) [inline sse2neon.h:7601]: I think fixed this line with the following code can solve the issue
```diff
-uint64x2_t result = vandq_u64(zf, cf);
+uint64x2_t result = vreinterpretq_u64_u32(vmvnq_u32(vreinterpretq_u32_u64(vorrq_u64(zf, cf))));
```

File: tests/impl.cpp [hunk 3]
```diff
@@ -9228,14 +9228,15 @@ result_t test_mm_test_mix_ones_zeros(const SSE2NEONTestImpl &impl,
...
-    int32_t d2 = !((_a[2]) & _mask[2]) & !((!_a[2]) & _mask[2]);
-    int32_t d3 = !((_a[3]) & _mask[3]) & !((!_a[3]) & _mask[3]);
-    int32_t result = ((d0 & d1 & d2 & d3) == 0) ? 1 : 0;
+    int32_t has_ones = 0;
+    int32_t has_zeros = 0;
+    for (int i = 0; i < 4; i++) {
+        has_ones |= _a[i] & _mask[i];
+        has_zeros |= ~_a[i] & _mask[i];
+    }
+    int32_t result = ((has_ones != 0) & (has_zeros != 0));
```
Comment (howjmay) [inline tests/impl.cpp:9237]: If possible maybe we can add some edge cases here to test the behavior?

File: sse2neon.h (hunk 1 above)
Comment (howjmay) [reply sse2neon.h]: I think we could just expand the implementation or use the fix I suggested in the following

File: sse2neon.h (hunk 2 above)
Comment (aqrit) [reply sse2neon.h:7601]: This suffers from the same problem. We can't `OR`/`AND` zf and cf together, until we've normalized then somehow. I'll see what I can do.

Comment (aqrit) [reply sse2neon.h:7601]: something like this for ARMv7-A ?
```
FORCE_INLINE int _mm_test_mix_ones_zeros(__m128i a, __m128i mask)
{
    uint64x2_t v = vreinterpretq_u64_m128i(a);
    uint64x2_t m = vreinterpretq_u64_m128i(mask);
    uint32x2_t has_one = vqmovn_u64(vandq_u64(m, v));
    uint32x2_t has_zero = vqmovn_u64(vbicq_u64(m, v));
    uint32x2_t pack = vpmax_u32(has_one, has_zero);
    uint32x2_t reduced = vpmin_u32(pack, pack);
    return (vget_lane_u32(reduced, 0) != 0);
}
```

Comment (aqrit) [reply sse2neon.h:7601]: added. remeber to squash it.

File: tests/impl.cpp (hunk 3 above)
Comment (aqrit) [reply tests/impl.cpp]: I'm not sure where the values are coming from, I'll leave this to you.

Comment (jserv) [reply tests/impl.cpp]: > If possible maybe we can add some edge cases here to test the behavior?

@howjmay, Can you show relevant code accordingly?

**Comment by jserv** [comment]
@aqrit, Additionally, I would like to acknowledge your significant contributions, including those to [sse2zig](https://github.com/aqrit/sse2zig), in the sse2neon header. Could you please let me know the preferred name or manner in which you'd like to be credited?

File: sse2neon.h [hunk 4]
```diff
@@ -7592,14 +7592,22 @@ FORCE_INLINE int _mm_test_all_zeros(__m128i a, __m128i mask)
 // zero, otherwise set CF to 0. Return 1 if both the ZF and CF values are zero,
 // otherwise return 0.
 // https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=mm_test_mix_ones_zero
+// Note: Argument names may be wrong in the Intel intrinsics guide.
```
Comment (howjmay) [inline sse2neon.h:7595]: One question about this. Why do you think the argument may be wrong. Just curious about the reason

Comment (aqrit) [reply sse2neon.h:7595]: according to the Release notes:

on `01/24/2023` 
`Corrected parameter order of _mm_test_all_zeros and _mm_test_mix_ones_zeros.`

I don't know if this person still maintains this or not:
https://old.reddit.com/r/simd/comments/cck1kh/feedback_on_intel_intrinsics_guide/

I don't have a twitter account so I can't really track them down.

File: tests/impl.cpp (hunk 3 above)
Comment (howjmay) [reply tests/impl.cpp]: I think if there is a test case all the bits of `mask` is zero would be good. It is a case that force the answer to be zero

File: sse2neon.h (hunk 4 above)
Comment (howjmay) [reply sse2neon.h:7595]: got it!! Thank you for explaining!!!

**Comment by jserv** [comment]
Thank @aqrit for contributing. It is time to make a new release.
